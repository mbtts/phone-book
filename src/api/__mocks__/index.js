const mockData = [
  {
    id: 0,
    name: "Oleta Level",
    phone_number: "+442032960159",
    address: "10 London Wall, London EC2M 6SA, UK"
  },
  {
    id: 6,
    name: "Stanley Vanderhoof",
    phone_number: "+442032960000",
    address: "17 Anchor Ave, Darwen BB3 0AZ, UK"
  }
];

const api = jest.fn().mockImplementation(() => {
  return Promise.resolve(mockData);
});

export { api as default, mockData };
