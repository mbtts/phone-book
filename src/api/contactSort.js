const createLastName = contact => {
  const lastName = contact.name.match(/\S+/g).pop();
  return lastName;
};

const sortByLastName = modifier => {
  return (a, b) =>
    createLastName(a).localeCompare(createLastName(b)) * modifier;
};

const sortByLastNameAsc = sortByLastName(1);
const sortByLastNameDesc = sortByLastName(-1);

export { sortByLastNameAsc, sortByLastNameDesc };
