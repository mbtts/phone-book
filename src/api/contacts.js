const api = async () => {
  const result = await fetch(
    "https://www.mocky.io/v2/581335f71000004204abaf83"
  );
  const json = await result.json();
  const indexed = json.contacts.map((contact, i) => ({ id: i, ...contact }));
  return indexed.sort(sortByLastNameAsc);
};

const createLastName = contact => {
  const lastName = contact.name.match(/\S+/g).pop();
  return lastName;
};

const sortByLastName = modifier => {
  return (a, b) => {
    return createLastName(a).localeCompare(createLastName(b)) * modifier;
  };
};

const sortByLastNameAsc = sortByLastName(1);
const sortByLastNameDesc = sortByLastName(-1);

export { api as default, sortByLastNameAsc, sortByLastNameDesc };
