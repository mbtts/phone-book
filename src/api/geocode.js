const GLOBAL_INIT = "initMaps";

const cache = {};

const processResults = (address, results, resolve, reject) => {
  if (results[0]) {
    const location = results[0].geometry.location;
    const coords = { lat: location.lat(), lng: location.lng() };
    cache[address] = coords;
    resolve(coords);
  } else {
    reject("No results found");
  }
};

const geoCode = address => {
  return new Promise((resolve, reject) => {
    if (cache[address]) {
      return resolve(cache[address]);
    }

    const geocoder = new google.maps.Geocoder();
    geocoder.geocode({ address }, (results, status) => {
      status === "OK"
        ? processResults(address, results, resolve, reject)
        : reject(`Geocode error ${status}`);
    });
  });
};

const createInitialiser = resolve => () => {
  window[GLOBAL_INIT] = undefined;
  resolve(google);
};

const initialiseMapsApi = () => {
  return new Promise(resolve => {
    if (typeof google !== "undefined") {
      return resolve(google);
    }

    window[GLOBAL_INIT] = createInitialiser(resolve);
    const script = document.createElement("script");
    script.setAttribute(
      "src",
      `https://maps.googleapis.com/maps/api/js?key=${
        process.env.GOOGLE_MAPS_API
      }&callback=${GLOBAL_INIT}`
    );
    document.body.appendChild(script);
  });
};

export { geoCode, initialiseMapsApi };
