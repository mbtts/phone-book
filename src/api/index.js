const api = async () => {
  const result = await fetch("http://www.mocky.io/v2/581335f71000004204abaf83");
  const json = await result.json();
  return json.contacts;
};

export default api;
