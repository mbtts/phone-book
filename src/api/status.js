const status = {
  LOADING: "LOADING",
  LOADED: "LOADED",
  ERROR: "ERROR"
};

export default status;
