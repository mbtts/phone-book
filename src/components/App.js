import { BrowserRouter, Route, Switch } from "react-router-dom";
import { ERROR, LOADED, LOADING } from "../api/status";
import React, { Component, Fragment } from "react";
import api, { sortByLastNameAsc, sortByLastNameDesc } from "../api";

import ContactList from "./ContactList";
import Detail from "./Detail";
import NoMatch from "./NoMatch";
import Repository from "../repository";
import Search from "./Search";
import SortButton from "./SortButton";
import sortByAlpha from "../../assets/svg/round-sort_by_alpha-24px.svg";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      status: LOADING,
      search: "",
      sortOrder: "asc"
    };
  }

  async componentDidMount() {
    try {
      const items = await api();
      this.setState(oldState => ({
        ...oldState,
        status: LOADED,
        items: new Repository(items)
      }));
    } catch (e) {
      this.setState(oldState => ({
        ...oldState,
        status: ERROR
      }));
    }
  }

  onSearch = e => {
    const search = e.target.value;

    this.setState(oldState => ({
      ...oldState,
      search,
      items: oldState.items.filter(item =>
        item.name.toLocaleLowerCase().includes(search.toLocaleLowerCase())
      )
    }));
  };

  onSortOrder = () => {
    this.setState(oldState => {
      const asc = oldState.sortOrder === "asc";
      const sortOrder = asc ? "desc" : "asc";
      const sortFn = asc ? sortByLastNameDesc : sortByLastNameAsc;
      return {
        ...oldState,
        sortOrder,
        items: oldState.items.sort(sortFn)
      };
    });
  };

  findItem = match => {
    const id = parseInt(match.params.id);
    const repository =
      this.state.items && this.state.items.filter(item => item.id === id);

    return repository && repository.selection.length === 1
      ? repository.selection[0]
      : null;
  };

  render() {
    const { status, items } = this.state;

    return (
      <Switch>
        <Route
          path="/"
          exact
          render={props => (
            <div className="contactlist">
              <div className="toolbar">
                <div className="toolbar__panel toolbar__panel--search">
                  <Search
                    value={this.state.search}
                    onChange={this.onSearch}
                    disabled={status !== LOADED}
                    className="toolbar__search"
                  />
                </div>
                <div className="toolbar__panel toolbar__panel--buttons">
                  <SortButton
                    icon={sortByAlpha}
                    order={this.state.sortOrder}
                    onClick={this.onSortOrder}
                    className="toolbar__button toolbar__button--wide"
                  />
                </div>
              </div>
              <ContactList status={status} repository={items} {...props} />
            </div>
          )}
        />
        <Route
          path="/user/:id(\d+)"
          render={props => (
            <Detail
              status={status}
              contact={this.findItem(props.match)}
              {...props}
            />
          )}
        />
        <Route component={NoMatch} />
      </Switch>
    );
  }
}

export default App;
