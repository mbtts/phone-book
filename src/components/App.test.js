import App from "./App";
import { LOADING } from "../api/status";
import { MemoryRouter } from "react-router-dom";
import React from "react";
import api from "../api";

jest.mock("../api");

const mockData = [
  {
    name: "Oleta Level",
    phone_number: "+442032960159",
    address: "10 London Wall, London EC2M 6SA, UK"
  }
];

describe("Phone book", () => {
  it("should render initial state", () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should be loading initially", () => {
    const wrapper = shallow(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    expect(wrapper.state("status")).toEqual(LOADING);
  });

  it("should load the contacts successfully", () => {
    api.mockReturnValueOnce(mockData);

    const wrapper = mount(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );
    return Promise.resolve().then(() => {
      wrapper.update();
      const p = wrapper.find("p");
      expect(p).toHaveLength(1);
      expect(p.text()).toMatch(mockData[0].name);
    });
  });

  it("should display an error if the contacts cannot load", () => {
    api.mockImplementationOnce(() => {
      throw new Error();
    });

    const wrapper = mount(<App />);
    return Promise.resolve().then(() => {
      wrapper.update();
      expect(wrapper.find("p").text()).toMatch("Error");
    });
  });
});
