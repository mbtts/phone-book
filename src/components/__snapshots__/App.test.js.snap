// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone book should render initial state 1`] = `
<div>
  <Search
    disabled={true}
    onChange={[Function]}
  />
  <p>
    Loadingâ€¦
  </p>
</div>
`;
